### PR TITLE
Added RPC command line options

### PR DESCRIPTION
**Pull request by andrasfuchs**
Added the following command line options
-rpcconnect -rpcport -rpcuser -rpcpassword

Added project profiles to test REST API and RPC settings

### CODE REVIEW

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 1]
```diff
@@ -8,72 +9,144 @@
...
+using Stratis.Bitcoin.Utilities.Extensions;
 
 namespace Stratis.Bitcoin.Cli
 {
     class Program
     {
         /// <summary>
         /// The expected sequence of arguments:
-        ///    1) [network-name] [rpc-command] [rpc-params].
+        ///    1) [network-name] [options] [rpc-command] [rpc-params].
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Hi @andrasfuchs! Ok so here rather use ```<list>``` and ```<item>```:

```
/// <summary>
/// The expected sequence of arguments:
/// <list>
/// <item> 1) [network-name] [options] [rpc-command] [rpc-params].

````

Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Hi @andrasfuchs! Ok so here rather use ```<list>``` and ```<item>```:

```
/// <summary>
/// The expected sequence of arguments:
/// <list>
/// <item> 1) [network-name] [options] [rpc-command] [rpc-params].

````

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 2]
```diff
@@ -8,72 +9,144 @@
...
         /// OR
-        ///    2) [network-name] [api-controller "/" api-command] [api-params].
+        ///    2) [network-name] [options] [api-controller "/" api-command] [api-params].
         /// </summary>
         public static void Main(string[] args)
         {
             try
             {
+                // Preprocess the command line arguments
+                List<string> argList = new List<string>(args);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: You can just use ```var``` here if the type is obvious.

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 3]
```diff
@@ -8,72 +9,144 @@
...
+
+                string networkName = null;
+                if (argList.Count > 0)
+                {
+                    networkName = argList.First();
+                    argList.RemoveAt(0);
+                }
+
+                List<string> optionList = new List<string>();
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: The same here, use ```var```

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 4]
```diff
@@ -8,72 +9,144 @@
...
+                }
+
+                string command = null;
+                if (argList.Count > 0)
+                {
+                    command = argList.First();
+                    argList.RemoveAt(0);
+                }
+
+                List<string> commandArgList = new List<string>(argList);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: The same here, use ```var```

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 5]
```diff
@@ -8,72 +9,144 @@
...
         /// </summary>
         public static void Main(string[] args)
         {
             try
             {
+                // Preprocess the command line arguments
+                List<string> argList = new List<string>(args);
+
+                string networkName = null;
+                if (argList.Count > 0)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: ```.Any()``` is better

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 6]
```diff
@@ -8,72 +9,144 @@
...
+
+                List<string> optionList = new List<string>();
+                while ((argList.Count > 0) && (argList[0].StartsWith('-')))
+                {
+                    optionList.Add(argList[0]);
+                    argList.RemoveAt(0);
+                }
+
+                string command = null;
+                if (argList.Count > 0)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: ```.Any()``` is better

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 7]
```diff
@@ -8,72 +9,144 @@
...
+                {
+                    command = argList.First();
+                    argList.RemoveAt(0);
+                }
+
+                List<string> commandArgList = new List<string>(argList);
+
                 // Display help if required.
-                if (args.Length == 0 || args.Contains("-help") || args.Contains("--help"))
+                if (command == null || argList.Contains("-help") || argList.Contains("--help"))
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Rather use ```string.IsNullOrWhiteSpace```

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 8]
```diff
@@ -8,72 +9,144 @@
...
+                    builder.AppendLine("dotnet run bitcoin getbalance -rpcuser=btctestuser -rpcpassword=btctestpass - Displays the current balance of the opened wallet on the 127.0.0.1:8332 node, authenticating with the RPC specified user.");
                     Console.WriteLine(builder);
                     return;
                 }
 
                 // Determine API port.
-                string blockchain = "bitcoin";
-                int apiPort = 37220;
-                if (args.Any(a => a.Contains("stratis")))
+                string blockchain = "";
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Rather use ```string.Empty``` or null

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 9]
```diff
@@ -8,72 +9,144 @@
...
-                if (!args.ElementAt(1).Contains("/"))
+                if (!command.Contains("/"))
                 {
                     // Process RPC call.
                     try
                     {
-                        NodeSettings nodeSettings = new NodeSettings(blockchain, network).LoadArguments(args);
+                        string[] options = optionList.ToArray();
+
+                        NodeSettings nodeSettings = new NodeSettings(blockchain, network).LoadArguments(options);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Can just use ```var```

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 10]
```diff
@@ -8,72 +9,144 @@
...
+                            rpcURI = new Uri($"http://{rpcAddress}:{rpcPort}");
+                        }
+                        rpcSettings.RpcUser = options.GetValueOf("-rpcuser") ?? rpcSettings.RpcUser;
+                        rpcSettings.RpcPassword = options.GetValueOf("-rpcpassword") ?? rpcSettings.RpcPassword;
+
+                        Console.WriteLine($"Connecting to the following RPC node: http://{rpcSettings.RpcUser}:{rpcSettings.RpcPassword}@{rpcURI.Authority}...");
 
-                        // Initialize the RPC client with the configured or passed userid, password and endpoint
-                        RPCClient rpc = new RPCClient($"{rpcSettings.RpcUser}:{rpcSettings.RpcPassword}", new Uri($"http://{nodeEndPoint}"));
+                        // Initilize the RPC client with the configured or passed userid, password and endpoint.
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Initialize is spelt wrong ;)

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 11]
```diff
@@ -8,72 +9,144 @@
...
+                        RPCClient rpc = new RPCClient($"{rpcSettings.RpcUser}:{rpcSettings.RpcPassword}", rpcURI, network);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Can use ```var```

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 1 above)
Comment (andrasfuchs) [reply src/Stratis.Bitcoin.Cli/Program.cs]: Hi @fassadlr , thank you for taking the time for the code review!
Sure, I didn't know about this summary tag.

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 5 above)
Comment (andrasfuchs) [reply src/Stratis.Bitcoin.Cli/Program.cs]: Is there any difference between the two?

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Cli/Program.cs]: @andrasfuchs just remember to change these too ;)

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 12]
```diff
@@ -57,7 +57,7 @@ public static void Main(string[] args)
                 var commandArgList = new List<string>(argList);
 
                 // Display help if required.
-                if (string.IsNullOrWhiteSpace(command) || argList.Contains("-help") || argList.Contains("--help"))
+                if (argList.Contains("-help") || argList.Contains("--help") || string.IsNullOrWhiteSpace(command))
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: ah sorry, what I meant was that this if{} should be the first thing in the method.

**Pull request merged**